cmip6_downscaling/methods/gard.py

```diff
@@ -0,0 +1,258 @@
+    seasonality = (
+        data.rolling({'time': seasonality_period}, center=True, min_periods=1)
+        .mean()
+        .groupby('time.dayofyear')
+        .mean()
+    )
+    detrended = data.groupby("time.dayofyear") - seasonality
+
+    # find spatial length scale
+    fields = detrended.values
+    bin_center, gamma = gs.vario_estimate(
+        # TODO: need to figure whether we need to do .values for this function
+        pos=(detrended.lon.values, detrended.lat.values),
+        field=fields,
+        latlon=True,
+        mesh_type='structured',
+    )
+    spatial = gs.Gaussian(dim=2, latlon=True, rescale=gs.EARTH_RADIUS)
+    spatial.fit_variogram(bin_center, gamma, sill=np.mean(np.var(fields, axis=(1, 2))))
+
+    # find temporal length scale
+    # break the time series into fields of 1 year length, since otherwise the algorithm struggles to find the correct length scale
+    fields = []
+    day_in_year = 365
+    for yr, group in detrended.groupby('time.year'):
+        # TODO: this is a very long list for a large domain, perhaps need random sampling
+        v = (
+            group.isel(time=slice(0, day_in_year))
+            .stack(point=['lat', 'lon'])
+            .transpose('point', 'time')
+            .values
+        )
+        fields.extend(list(v))
+    t = np.arange(day_in_year) / temporal_scaler
+    bin_center, gamma = gs.vario_estimate(pos=t, field=fields, mesh_type='structured')
+    temporal = gs.Gaussian(dim=1)
+    temporal.fit_variogram(bin_center, gamma, sill=np.mean(np.var(fields, axis=1)))
+
+    return {'temporal': temporal.len_scale, 'spatial': spatial.len_scale}
+
+
+def generate_scrf(
+    data: xr.Dataset,
+    label: str,
+    n_timepoints: int = 10958,
+    seasonality_period: int = 31,
+    seed: int = 0,
+    temporal_scaler: float = 1000.0,
+    crs: str = 'ESRI:54008',
+    **kwargs,
+) -> xr.DataArray:
+    """
+
+
+    Parameters
+    ----------
+    data : xr.Dataset
+        Data used to find the spatial and temporal correlation lengths
+    label : str
+        Name of the variable to be predicted
+    n_timepoints : int, optional
+        Number of timepoints to return
+    seasonality_period: int = 31,
+    seed: int = 0,
+    temporal_scaler: float = 1000.0,
+    crs: str = 'ESRI:54008',
+
+    """
+    # find correlation length from source data
+    length_scale = calc_correlation_length_scale(
+        data=data[label].isel(lat=slice(100, 110), lon=slice(100, 110)),
+        seasonality_period=seasonality_period,
+        temporal_scaler=temporal_scaler,
+    )
+    ss = length_scale['spatial']
+    ts = length_scale['temporal']
+
+    # reproject template into Sinusoidal projection
+    # TODO: any better ones for preserving distance between two arbitrary points on the map?
+    if 'x' not in data[label].dims:
+        template = data[label].rename({'lon': 'x', 'lat': 'y'})
+    projected = template.isel(time=0).rio.write_crs('EPSG:4326').rio.reproject(crs)
```

Comment:
`4326` is the lat/lon grid, so I don't think it would be changed. I was wondering if there's a better one to be used in place of the default for `crs` though (currently using the sinusoidal projection, see the TODO)